cardano-wallet server on _any_ arbitrary testnet

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

N/A

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- e0f5ef5aea50d01da290e53c795da4572303171c
  fix mainnet genesis hash
  
- dc0048d8536d0000bf5891512b2dd68270a625e6
  embed network magic inside Testnet discriminant
  The rationale is that, for mainnet, it is both known and not actually
required (addresses don't have protocol magic attributes). But for testnets
(plural), we may want an easy way to switch from a network to another.

- 22b914ff71331a40ed0fd9e644d976757f45ee7b
  allow passing custom testnets configuration via command-line
    e.g.

  ```
  $ cardano-wallet serve --mainnet \
    --node-socket ../cardano-node/state-node-mainnet/node.socket
  ```

  ```
  $ cardano-wallet serve --testnet 1097911063 \
    --genesis /nix/store/7106qjwral616j6p5049m7hhi5ih0r33-testnet-genesis.json \
    --node-socket ../cardano-node/state-node-testnet/node.socket
  ```

- 70cd8be7fe5875e65f9a86b6651d6bcacce41a32
  fixup fromGenesisTxOut
  
- 7fc880d51a1eb0f0f87cf83ca3a55472022f94c9
  adjust NetworkDiscriminant in test code to cope with recent change
  
- 7b9727a1d538c9089b18ec63a7411ce132901073
  remove hard-coded references to network discriminant from integration scenarios
  And rely instead on class constraints. There's no reason to specialize the network discriminant too early. The only moment
where we would want to do that is when calling 'serveWallet'.

- 896bf0f134ea002ec7fe0fd1e94c0ee944388bcc
  re-generate nix machinery
  
- ab155f7ae9a663d941291efa0b50c3479a1d8083
  improve error message when failing to parse genesis configuration
  




# Comments

<!-- Additional comments or screenshots to attach if any -->

This last commit is really not interesting, sorry for this. It is simply adjusting the network discriminant in many places. Note that I have switched from testnet to mainnet in a few places because it's easier to use and doesn't quite matter most of the time.

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ committed:
fix mainnet genesis hash

KtorZ committed:
embed network magic inside Testnet discriminant

The rationale is that, for mainnet, it is both known and not actually
required (addresses don't have protocol magic attributes). But for testnets
(plural), we may want an easy way to switch from a network to another.

KtorZ committed:
allow passing custom testnets configuration via command-line

  e.g.

  ```
  $ cardano-wallet serve --mainnet \
    --node-socket ../cardano-node/state-node-mainnet/node.socket
  ```

  ```
  $ cardano-wallet serve --testnet 1097911063 \
    --genesis /nix/store/7106qjwral616j6p5049m7hhi5ih0r33-testnet-genesis.json \
    --node-socket ../cardano-node/state-node-testnet/node.socket
  ```

KtorZ committed:
fixup fromGenesisTxOut

KtorZ committed:
adjust NetworkDiscriminant in test code to cope with recent change

rvl commented on lib/byron/exe/cardano-wallet-byron.hs:
```diff
@@ -176,37 +192,46 @@ cmdServe = command "serve" $ info (helper <*> helper' <*> cmd) $ mempty
+                , emptyGenesis (blockchainParameters @'Mainnet)
+                )
+            TestnetConfig (discriminant, vData) genesisFile -> do
+                (genesisData, genesisHash) <- unsafeRunExceptT $
```
Could add a `-- TODO: Error handling` here.

KtorZ replied:
Ho yes! I actually have it on my TODO list and will fix before the PR is merged. Passing a wrong genesis file results in a very ugly error at the moment ...

KtorZ replied:
Fixed:

```
$ cardano-wallet-byron serve --testnet 1234 --genesis shell.nix  --node-socket .travis.yml 
[cardano-wallet.main:Info:4] [2020-03-16 17:19:22.58 UTC] Running as v2020.3.16 (git revision: 8768ab1a1ab13af5aa9739c47ded3c995121fabe)
[cardano-wallet.main:Info:4] [2020-03-16 17:19:22.58 UTC] Command line: /home/ktorz/.local/bin/cardano-wallet-byron serve --testnet 1234 --genesis shell.nix --node-socket .travis.yml
[cardano-wallet.main:Debug:4] [2020-03-16 17:19:22.58 UTC] ServeArgs {_hostPreference = Host "127.0.0.1", _listen = ListenOnPort 8090, _nodeSocket = ".travis.yml", _networkConfiguration = TestnetConfig "shell.nix", _database = Nothing, _syncTolerance = SyncTolerance 300s, _enableShutdownHandler = False, _logging = LoggingOptions {loggingMinSeverity = Debug, loggingTracers = Tracers {applicationTracer = Const (Just Info), apiServerTracer = Const (Just Info), walletEngineTracer = Const (Just Info), walletDbTracer = Const (Just Info), ntpClientTracer = Const (Just Info)}, loggingTracersDoc = Nothing}}
[cardano-wallet.main:Error:4] [2020-03-16 17:19:22.58 UTC] Failed to parse genesis configuration. You may want to check the filepath given via --genesis and make sure it points to a valid JSON genesis file. Here's (perhaps) some helpful hint: GenesisDataParseError "(line 1, column 1):\nunexpected \"(\"\nexpecting white space or JSON value"
[cardano-wallet.main:Debug:4] [2020-03-16 17:19:22.58 UTC] Logging shutdown.
```

rvl reviewed (approved):
Looks good.

I think this wouldn't handle the case where the node is in a dev cluster with network discrimination disabled and its own genesis. But that probably doesn't matter.

paweljakubas commented on lib/byron/exe/cardano-wallet-byron.hs:
```diff
@@ -220,18 +245,68 @@ cmdServe = command "serve" $ info (helper <*> helper' <*> cmd) $ mempty
+                    ( SomeNetworkDiscriminant $ mapProxy proxy
+                    , testnetVersionData pm
+                    )
+                _ -> error "networkDiscriminantFlag: impossible: failed to \
```
maybe `impossible` was in the earlier version and forgot to erase? `failed to convert ProtocolMagic to SomeNat` is not enough?

KtorZ replied:
fixed.

paweljakubas commented on lib/byron/exe/cardano-wallet-byron.hs:
```diff
@@ -220,18 +245,68 @@ cmdServe = command "serve" $ info (helper <*> helper' <*> cmd) $ mempty
+        -> FilePath
+        -> NetworkConfiguration
+
+-- | Hand-written as there's no Show instance for 'NodeVersionData'
```
I wonder why Show and not ToText here (it is needed in `MsgNetworkName Text`) ?

KtorZ replied:
That's a good question. I will dig a bit more about why `Show` is needed for `MsgNeworkName`

paweljakubas reviewed (approved):
Looks fine (modulo nix voodoo - I am also applying patch from CI rather than producing locally;-)). General comment. Not huge change induced massive number of changes throughout the code. I wonder if we could somehow limit this range...

KtorZ committed:
remove hard-coded references to network discriminant from integration scenarios

And rely instead on class constraints. There's no reason to specialize the network discriminant too early. The only moment
where we would want to do that is when calling 'serveWallet'.

KtorZ committed:
re-generate nix machinery

KtorZ committed:
improve error message when failing to parse genesis configuration
